src/frontend/mame/ui/state.cpp

```diff
@@ -0,0 +1,424 @@
+// license:BSD-3-Clause
+// copyright-holders:Nathan Woods
+/***************************************************************************
+
+    ui/state.cpp
+
+    Menus for saving and loading state
+
+***************************************************************************/
+
+#include "emu.h"
+#include "machine.h"
+#include "emuopts.h"
+#include "ui/state.h"
+
+
+/***************************************************************************
+    ANONYMOUS NAMESPACE
+***************************************************************************/
+
+namespace {
+
+const int MAX_SAVED_STATE_JOYSTICK = 4;
+
+//-------------------------------------------------
+//  keyboard_code
+//-------------------------------------------------
+
+input_code keyboard_code(input_item_id id)
+{
+	// only supported for A-Z|0-9
+	assert((id >= ITEM_ID_A && id <= ITEM_ID_Z) || (id >= ITEM_ID_0 && id <= ITEM_ID_9));
+	return input_code(DEVICE_CLASS_KEYBOARD, 0, ITEM_CLASS_SWITCH, ITEM_MODIFIER_NONE, id);
+}
+
+
+//-------------------------------------------------
+//  keyboard_input_item_name
+//-------------------------------------------------
+
+std::string keyboard_input_item_name(input_item_id id)
+{
+	if (id >= ITEM_ID_A && id <= ITEM_ID_Z)
+		return std::string(1, (char)(id - ITEM_ID_A + 'a'));
```

Comment:
Can you drop the extra parens adding clutter?  This is C++ - `char(a  - b + c)` works just fine.